src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java

```diff
@@ -129,7 +129,8 @@ public void execute(
             boolean restored = result.isSuccess(); // if partially restored need to save increment
             if (restorable) {
                 restored &= restoreProject(result, mojoExecutions, mojoExecutionRunner, cacheConfig);
-            } else {
+            }
+            if (!restored) {
```

Comment:
perhaps we need to execute all mojos in this case including for Clean phase to remove any "garbage" which could left after an unsuccessful restore

Reply:
Good idea @maximilian-novikov-db.

@AlexanderAshitkin : do you mean we should change the way we restore artefacts from remote caches? I'm more inclined to the "second clean phase" idea since it covers any restoration problem.

Reply:
That is discussable:
* `clean` phase has been completed by this time, and repeating it sounds slightly off to me. 
* Keeping partially restored artifacts is error-prone, and `clean` might not even be requested. 
* invoking ad-hoc `clean` might wipe out other cached data and might be undesirable

It is better to leave it as is and rely on the regular `clean`, or implement `mv` logic to minimize the risk of corrupted cache data in the target dir.

Reply:
> * Keeping partially restored artifacts is error-prone, and `clean` might not even be requested.

all mojos except clean are executed again if it falls back to the "normal" build, i'd say keeping any partially restored data is harmful and may lead to unpredictable result

Reply:
I meant that the `clean` might not be requested by the user. In that case, using the `clean` looks undesirable because it has side effects outside the cache that might confuse the user. And I agree that leaving partially restored data is error-prone. But all three options (leaving as is, adding clean, or moving) have drawbacks.

Reply:
i suggest to remove the filter here https://github.com/apache/maven-build-cache-extension/blob/4d53a0a05c1e006265e19bce8ba8c9a0eb7831df/src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java#L137

and not to add Clean if it's not requested by user

Reply:
I agree that seeing a second clean phase seems a bit "off" but I could be in piece with it since it is a abnormal edge case (and we will write a beautiful log to explain it. :p)

Trying to summarize the exchange in order to check if we have a consensus in a situation where the cache fails the restoration
- the build was requested without clean phase : we do not change the actual behaviour.
- the build was requested with a clean phase : an extra clean phase will be executed before the actual behaviour (= the whole build is executed).

We change generically the log "Cannot restore project artifacts, continuing with non cached build" by : 
"Cannot restore project artifacts, continuing with non cached build (clean phase will be re-executed if requested)".

And in another PR, the remote download could be enhanced by doing it in two steps : downloading in a temp directory + moving it
Reducing the risk of leftovers files.

Reply:
Some clarifications about the current implementation:
* Logic of rerunning on failures is already within the BuildCacheMojosExecutionStrategy#restoreProject:
```java
if (!restored) {
    // cleanup partial state
    project.getArtifact().setFile(null);
    project.getArtifact().setResolved(false);
    mojoListener.remove(project);
    // build as usual
    for (MojoExecution mojoExecution : cachedSegment) { <<< rebuilding everything
        mojoExecutionRunner.run(mojoExecution);
    }
}
```
* `!restored` condition should be read as "at least part of the project were rebuilt, not restored"
* It doesn't mean an error automatically, it means something was rebuilt (at least in the design). 100% successful completion of the `restoreProject` could return `false`. 

So, in all, it feels like we might need fixes within the `restoreProject`:
1) to implement proper clean phase
2) Add a verification step to the "forced" steps - to force cache update on the detected inconsistencies
3) Make sure that in case of additional executions ("post-cached" or similar), we return `false` to update the cache.

Reply:
Alright, finally had some time to work on what you pointed out : regular build after a failed try to use the cache was handled differently depending the situation.
I will add comments on some lines of this commit to improve your ability to review it effectively. But here are some global points : 
- I moved the part where we check cached mojos consistency before artifact restoration. There are two benefits : this steps is faster than restoration over the network. So if we can't use the cache because of mojos parameters, better to know it before downloading artifacts. And secondly: there are no more "things" to rollback since setting artifacts of the project is done at the very end of the restoration.
- I kept the regular build execution outside the "restoreProject" function.
- I execute an extra clean phase only if two conditions are met : clean phase was requested in the build + there are potentially files restored in the project directory (it does not happen if we don't restore generated resources, "outputArtifacts" or if an error occurs before)  

I'm removing the draft on this PR since I don't want anymore to link it with my other PR. The duplicated 3 line function used in IT test can be remove in a very small 3rd PR.